File: jupyter2/problems/minc/minc/go/minc/parse.go
package main

import "fmt"

type ParseError struct{ Msg string }

func (e *ParseError) Error() string { return "parse error: " + e.Msg }

type parser struct {
	toks []Token
	pos  int
}

func (p *parser) peek() Token { return p.toks[p.pos] }

func (p *parser) advance() Token {
	t := p.toks[p.pos]
	if p.pos < len(p.toks)-1 {
		p.pos++
	}
	return t
}

func (p *parser) expect(k TokKind) {
	t := p.peek()
	if t.Kind != k {
		panic(&ParseError{fmt.Sprintf("expected %s but got %s", Token{Kind: k}, t)})
	}
	p.advance()
}

// expr = equality_expr "=" expr | equality_expr
func (p *parser) parseExpr() Expr {
	lhs := p.parseEquality()
	if p.peek().Kind == TAssign {
		p.advance()
		return &Assign{LHS: lhs, RHS: p.parseExpr()}
	}
	return lhs
}

// equality_expr = cmp_expr { ("=="|"!=") cmp_expr }*
func (p *parser) parseEquality() Expr {
	lhs := p.parseCmp()
	for {
		switch p.peek().Kind {
		case TEq:
			p.advance()
			lhs = &Binary{OpEq, lhs, p.parseCmp()}
		case TNe:
			p.advance()
			lhs = &Binary{OpNe, lhs, p.parseCmp()}
		default:
			return lhs
		}
	}
}

// cmp_expr = additive_expr { ("<="|">="|"<"|">") additive_expr }*
func (p *parser) parseCmp() Expr {
	lhs := p.parseAdditive()
	for {
		var op BinaryOp
		switch p.peek().Kind {
		case TLt:
			op = OpLt
		case TGt:
			op = OpGt
		case TLe:
			op = OpLe
		case TGe:
			op = OpGe
		default:
			return lhs
		}
		p.advance()
		lhs = &Binary{op, lhs, p.parseAdditive()}
	}
}

// additive_expr = multiplicative_expr { ("+"|"-") multiplicative_expr }*
func (p *parser) parseAdditive() Expr {
	lhs := p.parseMul()
	for {
		var op BinaryOp
		switch p.peek().Kind {
		case TPlus:
			op = OpAdd
		case TMinus:
			op = OpSub
		default:
			return lhs
		}
		p.advance()
		lhs = &Binary{op, lhs, p.parseMul()}
	}
}

// multiplicative_expr = unary_expr { ("*"|"/"|"%") unary_expr }*
func (p *parser) parseMul() Expr {
	lhs := p.parseUnary()
	for {
		var op BinaryOp
		switch p.peek().Kind {
		case TStar:
			op = OpMul
		case TSlash:
			op = OpDiv
		case TPercent:
			op = OpMod
		default:
			return lhs
		}
		p.advance()
		lhs = &Binary{op, lhs, p.parseUnary()}
	}
}

// unary_expr = number | identifier "(" arg_list ")" | identifier | "(" expr ")" | ("+"|"-"|"!"|"~") unary_expr
func (p *parser) parseUnary() Expr {
	t := p.peek()
	switch t.Kind {
	case TNum:
		p.advance()
		return &IntLit{t.Num}
	case TPlus:
		p.advance()
		return &Unary{OpPos, p.parseUnary()}
	case TMinus:
		p.advance()
		return &Unary{OpNeg, p.parseUnary()}
	case TNot:
		p.advance()
		return &Unary{OpLNot, p.parseUnary()}
	case TTilde:
		p.advance()
		return &Unary{OpBNot, p.parseUnary()}
	case TLParen:
		p.advance()
		e := p.parseExpr()
		p.expect(TRParen)
		return e
	case TId:
		p.advance()
		if p.peek().Kind == TLParen {
			p.advance()
			args := p.parseArgList()
			p.expect(TRParen)
			return &Call{Name: t.Str, Args: args}
		}
		return &VarRef{t.Str}
	default:
		panic(&ParseError{fmt.Sprintf("unexpected token %s in expression", t)})
	}
}

// arg_list = [ expr { "," expr }* ]
func (p *parser) parseArgList() []Expr {
	if p.peek().Kind == TRParen {
		return nil
	}
	args := []Expr{p.parseExpr()}
	for p.peek().Kind == TComma {
		p.advance()
		args = append(args, p.parseExpr())
	}
	return args
}

// type_expr = "long"
func (p *parser) parseType() TypeExpr {
	p.expect(TLong)
	return TyLong
}

func (p *parser) parseIdent() string {
	t := p.advance()
	if t.Kind != TId {
		panic(&ParseError{fmt.Sprintf("expected identifier but got %s", t)})
	}
	return t.Str
}

// stmt = ";" | "continue" ";" | "break" ";" | "return" expr ";" | compound_stmt | if_stmt | while_stmt | expr ";"
func (p *parser) parseStmt() Stmt {
	switch p.peek().Kind {
	case TSemi:
		p.advance()
		return &EmptyStmt{}
	case TContinue:
		p.advance()
		p.expect(TSemi)
		return &ContinueStmt{}
	case TBreak:
		p.advance()
		p.expect(TSemi)
		return &BreakStmt{}
	case TReturn:
		p.advance()
		e := p.parseExpr()
		p.expect(TSemi)
		return &ReturnStmt{e}
	case TLBrace:
		return p.parseCompound()
	case TIf:
		return p.parseIf()
	default:
		e := p.parseExpr()
		p.expect(TSemi)
		return &ExprStmt{e}
	}
}

// compound_stmt = "{" {var_decl}* {stmt}* "}"
func (p *parser) parseCompound() *Compound {
	p.expect(TLBrace)
	var decls []VarDecl
	for p.peek().Kind == TLong {
		ty := p.parseType()
		name := p.parseIdent()
		p.expect(TSemi)
		decls = append(decls, VarDecl{ty, name})
	}
	var stmts []Stmt
	for p.peek().Kind != TRBrace {
		stmts = append(stmts, p.parseStmt())
	}
	p.expect(TRBrace)
	return &Compound{Decls: decls, Stmts: stmts}
}

// if_stmt = "if" "(" expr ")" stmt [ "else" stmt ]
func (p *parser) parseIf() Stmt {
	p.expect(TIf)
	p.expect(TLParen)
	cond := p.parseExpr()
	p.expect(TRParen)
	then := p.parseStmt()
	var els Stmt
	if p.peek().Kind == TElse {
		p.advance()
		els = p.parseStmt()
	}
	return &IfStmt{Cond: cond, Then: then, Else: els}
}


// parameter = type_expr identifier
func (p *parser) parseParam() Param {
	ty := p.parseType()
	return Param{Type: ty, Name: p.parseIdent()}
}

// parameter_list = [ parameter { "," parameter }* ]
func (p *parser) parseParamList() []Param {
	if p.peek().Kind == TRParen {
		return nil
	}
	params := []Param{p.parseParam()}
	for p.peek().Kind == TComma {
		p.advance()
		params = append(params, p.parseParam())
	}
	return params
}

// fun_definition = type_expr identifier "(" parameter_list ")" compound_stmt
func (p *parser) parseFunDef() *FunDef {
	ret := p.parseType()
	name := p.parseIdent()
	p.expect(TLParen)
	params := p.parseParamList()
	p.expect(TRParen)
	body := p.parseCompound()
	return &FunDef{RetType: ret, Name: name, Params: params, Body: body}
}

// program = {definition}*
func (p *parser) parseProgram() *Program {
	var funcs []*FunDef
	for p.peek().Kind != TEOF {
		funcs = append(funcs, p.parseFunDef())
	}
	return &Program{Funcs: funcs}
}

func Parse(toks []Token) (prog *Program, err error) {
	defer func() {
		if r := recover(); r != nil {
			if pe, ok := r.(*ParseError); ok {
				err = pe
				return
			}
			panic(r)
		}
	}()
	p := &parser{toks: toks}
	prog = p.parseProgram()
	if p.peek().Kind != TEOF {
		panic(&ParseError{fmt.Sprintf("trailing tokens, starting at %s", p.peek())})
	}
	return prog, nil
}

func ParseString(s string) (*Program, error) {
	toks, err := Tokenize(s)
	if err != nil {
		return nil, err
	}
	return Parse(toks)
}

File: jupyter2/problems/minc/minc/jl/minc/parse.jl
mutable struct TokenStream
    toks::Vector{Token}
    pos::Int
end

# report an error at the point it occurs and return nothing
function parsefail(msg::String)
    println(stderr, "parse error: " * msg)
    return nothing
end

peektok(t::TokenStream) = t.toks[t.pos]

function advance!(t::TokenStream)
    tok = t.toks[t.pos]
    if t.pos < length(t.toks)
        t.pos += 1
    end
    return tok
end

function expect!(t::TokenStream, k::TokKind)::Bool
    tok = peektok(t)
    if tok.kind != k
        parsefail("expected $(tokstr(Token(k))) but got $(tokstr(tok))")
        return false
    end
    advance!(t)
    return true
end

# expr = equality_expr "=" expr | equality_expr
function parseExpr(t::TokenStream)::Union{Expr,Nothing}
    lhs = parseEquality(t)
    lhs === nothing && return nothing
    if peektok(t).kind == TAssign
        advance!(t)
        rhs = parseExpr(t)
        rhs === nothing && return nothing
        return Assign(lhs, rhs)
    end
    return lhs
end

# equality_expr = cmp_expr { ("=="|"!=") cmp_expr }*
function parseEquality(t::TokenStream)::Union{Expr,Nothing}
    lhs = parseCmp(t)
    lhs === nothing && return nothing
    while true
        k = peektok(t).kind
        if k == TEq || k == TNe
            advance!(t)
            rhs = parseCmp(t)
            rhs === nothing && return nothing
            lhs = Binary(k == TEq ? OpEq : OpNe, lhs, rhs)
        else
            return lhs
        end
    end
end

# cmp_expr = additive_expr { ("<="|">="|"<"|">") additive_expr }*
function parseCmp(t::TokenStream)::Union{Expr,Nothing}
    lhs = parseAdditive(t)
    lhs === nothing && return nothing
    while true
        k = peektok(t).kind
        op = if k == TLt
            OpLt
        elseif k == TGt
            OpGt
        elseif k == TLe
            OpLe
        elseif k == TGe
            OpGe
        else
            return lhs
        end
        advance!(t)
        rhs = parseAdditive(t)
        rhs === nothing && return nothing
        lhs = Binary(op, lhs, rhs)
    end
end

# additive_expr = multiplicative_expr { ("+"|"-") multiplicative_expr }*
function parseAdditive(t::TokenStream)::Union{Expr,Nothing}
    lhs = parseMul(t)
    lhs === nothing && return nothing
    while true
        k = peektok(t).kind
        op = if k == TPlus
            OpAdd
        elseif k == TMinus
            OpSub
        else
            return lhs
        end
        advance!(t)
        rhs = parseMul(t)
        rhs === nothing && return nothing
        lhs = Binary(op, lhs, rhs)
    end
end

# multiplicative_expr = unary_expr { ("*"|"/"|"%") unary_expr }*
function parseMul(t::TokenStream)::Union{Expr,Nothing}
    lhs = parseUnary(t)
    lhs === nothing && return nothing
    while true
        k = peektok(t).kind
        op = if k == TStar
            OpMul
        elseif k == TSlash
            OpDiv
        elseif k == TPercent
            OpMod
        else
            return lhs
        end
        advance!(t)
        rhs = parseUnary(t)
        rhs === nothing && return nothing
        lhs = Binary(op, lhs, rhs)
    end
end

# unary_expr = number | identifier "(" arg_list ")" | identifier | "(" expr ")" | ("+"|"-"|"!"|"~") unary_expr
function parseUnary(t::TokenStream)::Union{Expr,Nothing}
    tok = peektok(t)
    k = tok.kind
    if k == TNum
        advance!(t)
        return IntLit(tok.num)
    elseif k == TPlus || k == TMinus || k == TNot || k == TTilde
        advance!(t)
        e = parseUnary(t)
        e === nothing && return nothing
        op = if k == TPlus
            OpPos
        elseif k == TMinus
            OpNeg
        elseif k == TNot
            OpLNot
        else
            OpBNot
        end
        return Unary(op, e)
    elseif k == TLParen
        advance!(t)
        e = parseExpr(t)
        e === nothing && return nothing
        expect!(t, TRParen) || return nothing
        return e
    elseif k == TId
        advance!(t)
        if peektok(t).kind == TLParen
            advance!(t)
            args = parseArgList(t)
            args === nothing && return nothing
            expect!(t, TRParen) || return nothing
            return Call(tok.str, args)
        end
        return VarRef(tok.str)
    else
        return parsefail("unexpected token $(tokstr(tok)) in expression")
    end
end

# arg_list = [ expr { "," expr }* ]
function parseArgList(t::TokenStream)::Union{Vector{Expr},Nothing}
    args = Expr[]
    if peektok(t).kind == TRParen
        return args
    end
    e = parseExpr(t)
    e === nothing && return nothing
    push!(args, e)
    while peektok(t).kind == TComma
        advance!(t)
        e = parseExpr(t)
        e === nothing && return nothing
        push!(args, e)
    end
    return args
end

# type_expr = "long"
function parseType(t::TokenStream)::Union{TypeExpr,Nothing}
    expect!(t, TLong) || return nothing
    return TyLong
end

# identifier = /[A-Za-z_][A-Za-z_0-9]*/
function parseIdent(t::TokenStream)::Union{String,Nothing}
    tok = advance!(t)
    if tok.kind != TId
        return parsefail("expected identifier but got $(tokstr(tok))")
    end
    return tok.str
end

# stmt = ";" | "continue" ";" | "break" ";" | "return" expr ";" | compound_stmt | if_stmt | while_stmt | expr ";"
function parseStmt(t::TokenStream)::Union{Stmt,Nothing}
    k = peektok(t).kind
    if k == TSemi
        advance!(t)
        return EmptyStmt()
    elseif k == TContinue
        advance!(t)
        expect!(t, TSemi) || return nothing
        return ContinueStmt()
    elseif k == TBreak
        advance!(t)
        expect!(t, TSemi) || return nothing
        return BreakStmt()
    elseif k == TReturn
        advance!(t)
        e = parseExpr(t)
        e === nothing && return nothing
        expect!(t, TSemi) || return nothing
        return ReturnStmt(e)
    elseif k == TLBrace
        return parseCompound(t)
    elseif k == TIf
        return parseIf(t)
    else
        e = parseExpr(t)
        e === nothing && return nothing
        expect!(t, TSemi) || return nothing
        return ExprStmt(e)
    end
end

# compound_stmt = "{" {var_decl}* {stmt}* "}"
function parseCompound(t::TokenStream)::Union{Compound,Nothing}
    expect!(t, TLBrace) || return nothing
    decls = VarDecl[]
    while peektok(t).kind == TLong
        ty = parseType(t)
        ty === nothing && return nothing
        name = parseIdent(t)
        name === nothing && return nothing
        expect!(t, TSemi) || return nothing
        push!(decls, VarDecl(ty, name))
    end
    stmts = Stmt[]
    while peektok(t).kind != TRBrace
        s = parseStmt(t)
        s === nothing && return nothing
        push!(stmts, s)
    end
    expect!(t, TRBrace) || return nothing
    return Compound(decls, stmts)
end

# if_stmt = "if" "(" expr ")" stmt [ "else" stmt ]
function parseIf(t::TokenStream)::Union{Stmt,Nothing}
    expect!(t, TIf) || return nothing
    expect!(t, TLParen) || return nothing
    cond = parseExpr(t)
    cond === nothing && return nothing
    expect!(t, TRParen) || return nothing
    then = parseStmt(t)
    then === nothing && return nothing
    els = nothing
    if peektok(t).kind == TElse
        advance!(t)
        els = parseStmt(t)
        els === nothing && return nothing
    end
    return IfStmt(cond, then, els)
end


# parameter = type_expr identifier
function parseParam(t::TokenStream)::Union{Param,Nothing}
    ty = parseType(t)
    ty === nothing && return nothing
    name = parseIdent(t)
    name === nothing && return nothing
    return Param(ty, name)
end

# parameter_list = [ parameter { "," parameter }* ]
function parseParamList(t::TokenStream)::Union{Vector{Param},Nothing}
    params = Param[]
    if peektok(t).kind == TRParen
        return params
    end
    pm = parseParam(t)
    pm === nothing && return nothing
    push!(params, pm)
    while peektok(t).kind == TComma
        advance!(t)
        pm = parseParam(t)
        pm === nothing && return nothing
        push!(params, pm)
    end
    return params
end

# fun_definition = type_expr identifier "(" parameter_list ")" compound_stmt
function parseFunDef(t::TokenStream)::Union{FunDef,Nothing}
    ret = parseType(t)
    ret === nothing && return nothing
    name = parseIdent(t)
    name === nothing && return nothing
    expect!(t, TLParen) || return nothing
    params = parseParamList(t)
    params === nothing && return nothing
    expect!(t, TRParen) || return nothing
    body = parseCompound(t)
    body === nothing && return nothing
    return FunDef(ret, name, params, body)
end

# program = {definition}*
function parseProgram(t::TokenStream)::Union{Program,Nothing}
    funcs = FunDef[]
    while peektok(t).kind != TEOF
        f = parseFunDef(t)
        f === nothing && return nothing
        push!(funcs, f)
    end
    return Program(funcs)
end

function parse_tokens(toks::Vector{Token})::Union{Program,Nothing}
    t = TokenStream(toks, 1)
    prog = parseProgram(t)
    prog === nothing && return nothing
    if peektok(t).kind != TEOF
        return parsefail("trailing tokens, starting at $(tokstr(peektok(t)))")
    end
    return prog
end

function parse_string(s::String)::Union{Program,Nothing}
    toks = tokenize(s)
    toks === nothing && return nothing
    return parse_tokens(toks)
end
